design/continuous-delivery.md

```diff
@@ -0,0 +1,578 @@
+
+For example, if we use CodePipeline for deploying an app to multiple environments,the deployment infrastructure will consist of a central pipeline stack, which contains the pipeline itself, it's artifacts bucket and other related resources such as CodeBuild projects. It will also require a stack in each region that includes a CodePipeline regional replication bucket (and key).
+
+In CodePipeline, we will implement this stage using a CodeBuild action which runs `cdk deploy "pipeline-*"`. This will deploy all stacks that begin with the `pipeline-` prefix. These stacks can be deployed to any bootstrapped environment since `cdk deploy` can assume the deployment role.
+
+
+## Publishing
+
+The [`cdk-assets`](./cdk-assets.md) tool is responsible for _packaging_ and _publishing_ application assets to "**asset stores**" in AWS environments so they can be consumed by stacks deployed to these environments.
+
+See the [cdk-assets specification](./cdk-assets.md) for additional details.
+
+```shell
+cdk-assets publish cdk.out [ASSET-ID,ASSET-ID,...]
+```
+
+The input is the *cloud assembly* (`cdk.out`), which includes an asset manifest `assets.json`:
+
+The manifest is synthesized by the app and, for each asset (identified by their source hash) includes the `source` information (within the cloud assembly) and as set of `destinations` with a list of locations into which the asset should be published. The idea is that this manifest is all the information the publish needs.
+
+A list of asset IDs (source hashes) can also be included in order to only publish a subset of the assets. This can be used to implement concurrent publishing of assets (e.g. through CodePipeline).
+
+Then, for each asset, cdk-assets will perform the following operation:
+
+1. Assume the **publishing IAM role** in the target environment.
+2. Check if the asset is already published to this location. Assets are identified by their source hash. If it is, skip.
+3. If the asset doesn’t exists locally (e.g. docker image already exists, zip file already exists in local cache), package (docker build, zip directory).
+4. Publish the asset to the target location.
+
+In order for the publish to be able to execute `docker build`, this command must be executed in an environment that has docker available (in CodeBuild this means the project must be "privileged").
+
+**Templates as Assets**
+
+Some deployment systems (e.g. the CDK CLI) require that CloudFormation templates will be uploaded an S3 location before they can be deployed (this is done automatically in CodePipeline). Due to S3 eventual consistency, these files must be immutable, so we need to upload a new template file every time the template changes.
+
+To that end, we will treat all CloudFormation templates in the assembly like any other asset. They will be identified by their source hash (the hash of the template) and uploaded to the asset store in the environment in which they are expected to be deployed, like any other file asset.
```

Comment:
But CodePipeline needs the templates to come from the synth artifact that contains `cdk.out`. What's the point then?

Reply:
For other systems like the Amazon internal systems.